app/gateway-oss/2.5.x/performance-testing-framework.md

```diff
@@ -0,0 +1,321 @@
+    end)
+  end)
+end
+```
+
+The test can be ran just as a normal busted based test. Run it with bin/busted path/to/this_file_spec.lua.
+
+More examples can be found at [spec/04-perf](https://github.com/Kong/kong/tree/master/spec/04-perf) of Kong
+respository.
+
+[Back to top](#introduction)
+
+## API
+
+### perf.use_driver
+
+*syntax: perf.use_driver(name, options?)*
+
+Use driver of name, which must be one of "local", "docker"  or  "terraform". Additional
+parameters for driver can be specified in options as a Lua table. Throws error if any.
+
+Only terraform driver expect options parameter, which contains following keys:
+
+- **provider** the service provider name, right now only "equinix-metal"
+- **tfvars** Terraform variables as a Lua table; for `equinix-metal` provider,
+`packet_project_id` and `packet_auth_token` is required.
+
+[Back to top](#introduction)
+
+### perf.set_log_level
+
+*syntax: perf.set_log_level(level)*
+
+Set the log level, expect one of `"debug"`, `"info"`, `"notice"`, `"warn"`, `"error"` and
+`"crit"`. The default log level is `"info"`.
+
+[Back to top](#introduction)
+
+### perf.set_retry_count
+
+*syntax: perf.set_retry_count(count)*
+
+Set retry time for each “driver" operation. By default every operation is retried 3 times.
+
+[Back to top](#introduction)
+
+### perf.setup
+
+*syntax: helpers = perf.setup()*
+
+Prepare environment and returns the spec.helpers module. Throws error if any.
+
+The framework setup some environment variables before importing spec.helpers modules.
+The returned helpers is just a nomal spec.helpers module, user can use same pattern
+in integration tests to setup entities in Kong. DB-less mode is currently not implemented.
+
+[Back to top](#introduction)
+
+### perf.start_upstream
+
+*syntax: upstream_uri = perf.start_upstream(nginx_conf_blob)*
+
+Start upstream (Nginx/OpenResty) with given nginx_conf_blob. Returns the upstream
+URI accessible from Kong's view. Throws error if any.
+
+[Back to top](#introduction)
+
+### perf.start_kong
+
+*syntax: perf.start_kong(version, kong_configs?)*
+
+Start Kong with given version and Kong configurations in kong_configs as a Lua table.
+Throws error if any.
+
+To select a git hash or tag, use `"git:<hash>"` as version. Otherwise, the framework
+will treat the `version` as a release version and will download binary packages from
+Kong's distribution website.
+
+[Back to top](#introduction)
+
+### perf.stop_kong
+
+*syntax: perf.stop_kong()*
+
+Stops Kong. Throws error if any.
+
+[Back to top](#introduction)
+
+### perf.teardown
+
+*syntax: perf.teardown(full?)*
+
+Teardown. Throws error if any. With terraform driver, set full to true terminates
+all infrastructure while by default it does cleanup only to speed up successive runs.
+
+[Back to top](#introduction)
+
+### perf.start_stapxx
+
+*syntax: perf.start_stapxx(stapxx_file_name, arg?)*
+
+Start the Stap++ script with stapxx_file_name exists in   and addtional CLI args args. Throws error if any.
```

Comment:
```suggestion
Starts the Stap++ script with `stapxx_file_name` exists in and additional CLI args. Throws error if any.
```

Reply:
Not sure what this means, "Start the Stap++ script with `stapxx_file_name` exists in and additional CLI args" - is something missing here?